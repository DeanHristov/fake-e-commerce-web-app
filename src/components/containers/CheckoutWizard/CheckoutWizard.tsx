'use client';

import { FC, useState } from 'react';

import WizardContent from './WizardContent/WizardContent';
import WizardHeading from './WizardHeading/WizardHeading';
import Widget from '@/components/containers/Widget';

export enum ACTIVE_VARIANT {
  PREVIEW = 1,
  DELIVERY = 2,
  RECEIPT = 3,
}

export interface ICheckoutWizardProps {}

const CheckoutWizard: FC<ICheckoutWizardProps> = () => {
  const [activeTab, setActiveTab] = useState<ACTIVE_VARIANT>(
    ACTIVE_VARIANT.PREVIEW,
  );

  return (
    <Widget
      title={'Check out'}
      className="w-full checkout-container min-h-full space-y-4"
    >
      <div className="tab-heading h-12 flex justify-center items-center">
        <WizardHeading
          variant={ACTIVE_VARIANT.PREVIEW}
          onClick={setActiveTab}
          active={activeTab}
        />
        <WizardHeading
          variant={ACTIVE_VARIANT.DELIVERY}
          onClick={setActiveTab}
          active={activeTab}
        />
        <WizardHeading
          variant={ACTIVE_VARIANT.RECEIPT}
          onClick={setActiveTab}
          active={activeTab}
        />
      </div>
      <div className="tab-content h-full flex">
        <WizardContent isActive={activeTab === 1}>
          <h3>{'// Add preview info'}</h3>
        </WizardContent>
        <WizardContent isActive={activeTab === 2}>
          <h3>{'// Add delivery info'}</h3>
        </WizardContent>
        <WizardContent isActive={activeTab === 3}>
          <h3>{'// Add receipt info'}</h3>
        </WizardContent>
      </div>
    </Widget>
  );
};

export default CheckoutWizard;
