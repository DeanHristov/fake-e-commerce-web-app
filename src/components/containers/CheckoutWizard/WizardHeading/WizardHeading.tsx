'use client';
import { FC } from 'react';
import {
  BuildingOffice2Icon,
  PencilSquareIcon,
  ShoppingBagIcon,
} from '@heroicons/react/24/solid';
import { ACTIVE_VARIANT } from '@/components/containers/CheckoutWizard/CheckoutWizard';

export interface IWizardHeadingProps {
  active: ACTIVE_VARIANT;
  onClick: (value: number) => void;
  variant: ACTIVE_VARIANT;
  disabled?: boolean;
}

const WizardHeading: FC<IWizardHeadingProps> = ({
  active,
  variant,
  onClick,
  disabled = false,
}) => {
  const isDisabled: boolean = !disabled && active === variant;

  return (
    <div
      className={`wizard-heading ${
        isDisabled ? 'border-green-500' : 'border-gray-300'
      }`}
    >
      <div
        data-testid={'circle-button'}
        onClick={() => !disabled && onClick(variant)}
        className={`wizard-heading-circle ${
          isDisabled ? 'border-green-500 active' : 'border-gray-300'
        }`}
      >
        {variant === ACTIVE_VARIANT.PREVIEW && (
          <ShoppingBagIcon
            className={`wizard-heading-icon ${
              isDisabled ? 'text-green-500' : 'text-gray-300'
            }`}
          />
        )}
        {variant === ACTIVE_VARIANT.DELIVERY && (
          <BuildingOffice2Icon
            className={`wizard-heading-icon ${
              isDisabled ? 'text-green-500' : 'text-gray-300'
            }`}
          />
        )}
        {variant === ACTIVE_VARIANT.RECEIPT && (
          <PencilSquareIcon
            className={`wizard-heading-icon ${
              isDisabled ? 'text-green-500' : 'text-gray-300'
            }`}
          />
        )}
      </div>
    </div>
  );
};

export default WizardHeading;
