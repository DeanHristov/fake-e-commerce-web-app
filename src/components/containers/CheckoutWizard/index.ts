import CheckoutWizard, {
  ACTIVE_VARIANT,
  ICheckoutWizardProps,
} from './CheckoutWizard';
import WizardHeading, {
  IWizardHeadingProps,
} from './WizardHeading/WizardHeading';
import WizardContent, {
  IWizardContentProps,
} from './WizardContent/WizardContent';

export type {
  ICheckoutWizardProps,
  IWizardHeadingProps,
  IWizardContentProps,
  ACTIVE_VARIANT,
};

export { WizardHeading, WizardContent };
export default CheckoutWizard;
