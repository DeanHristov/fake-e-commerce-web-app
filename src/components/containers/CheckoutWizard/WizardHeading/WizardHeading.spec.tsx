import { render } from '@testing-library/react';

import WizardHeading from './WizardHeading';
import { ACTIVE_VARIANT } from '@/components/containers/CheckoutWizard/CheckoutWizard';
import userEvent from '@testing-library/user-event';

describe('UI/Controls <WizardHeading {...} />', () => {
  it('Should be able to render the active state of the Component', () => {
    const { container } = render(
      <WizardHeading
        onClick={jest.fn}
        active={ACTIVE_VARIANT.RECEIPT}
        variant={ACTIVE_VARIANT.RECEIPT}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('Should be able to render the in-active state of the Component', () => {
    const { container } = render(
      <WizardHeading
        onClick={jest.fn}
        active={ACTIVE_VARIANT.PREVIEW}
        variant={ACTIVE_VARIANT.RECEIPT}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it.skip('Should be able to trigger a click event', async () => {
    const mockCB = jest.fn();
    const { getByTestId } = render(
      <WizardHeading
        onClick={mockCB}
        active={ACTIVE_VARIANT.PREVIEW}
        variant={ACTIVE_VARIANT.RECEIPT}
      />,
    );

    const $button: HTMLElement = getByTestId('circle-button');

    await userEvent.click($button);

    expect(mockCB).toHaveBeenCalled();
  });

  it('Should not be able to trigger a click event if it is disabled', async () => {
    const mockCB = jest.fn();
    const { container, getByTestId } = render(
      <WizardHeading
        disabled
        onClick={mockCB}
        active={ACTIVE_VARIANT.PREVIEW}
        variant={ACTIVE_VARIANT.RECEIPT}
      />,
    );

    const $button: HTMLElement = getByTestId('circle-button');

    await userEvent.click($button);

    expect(mockCB).not.toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });
});
