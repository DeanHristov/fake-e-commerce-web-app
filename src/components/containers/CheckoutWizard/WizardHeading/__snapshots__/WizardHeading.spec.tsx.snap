// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI/Controls <WizardHeading {...} /> Should be able to render the active state of the Component 1`] = `
<div>
  <div
    class="wizard-heading border-green-500"
  >
    <div
      class="wizard-heading-circle border-green-500 active"
      data-testid="circle-button"
    >
      <svg
        aria-hidden="true"
        class="wizard-heading-icon text-green-500"
        data-slot="icon"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21.731 2.269a2.625 2.625 0 0 0-3.712 0l-1.157 1.157 3.712 3.712 1.157-1.157a2.625 2.625 0 0 0 0-3.712ZM19.513 8.199l-3.712-3.712-8.4 8.4a5.25 5.25 0 0 0-1.32 2.214l-.8 2.685a.75.75 0 0 0 .933.933l2.685-.8a5.25 5.25 0 0 0 2.214-1.32l8.4-8.4Z"
        />
        <path
          d="M5.25 5.25a3 3 0 0 0-3 3v10.5a3 3 0 0 0 3 3h10.5a3 3 0 0 0 3-3V13.5a.75.75 0 0 0-1.5 0v5.25a1.5 1.5 0 0 1-1.5 1.5H5.25a1.5 1.5 0 0 1-1.5-1.5V8.25a1.5 1.5 0 0 1 1.5-1.5h5.25a.75.75 0 0 0 0-1.5H5.25Z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`UI/Controls <WizardHeading {...} /> Should be able to render the in-active state of the Component 1`] = `
<div>
  <div
    class="wizard-heading border-gray-300"
  >
    <div
      class="wizard-heading-circle border-gray-300"
      data-testid="circle-button"
    >
      <svg
        aria-hidden="true"
        class="wizard-heading-icon text-gray-300"
        data-slot="icon"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21.731 2.269a2.625 2.625 0 0 0-3.712 0l-1.157 1.157 3.712 3.712 1.157-1.157a2.625 2.625 0 0 0 0-3.712ZM19.513 8.199l-3.712-3.712-8.4 8.4a5.25 5.25 0 0 0-1.32 2.214l-.8 2.685a.75.75 0 0 0 .933.933l2.685-.8a5.25 5.25 0 0 0 2.214-1.32l8.4-8.4Z"
        />
        <path
          d="M5.25 5.25a3 3 0 0 0-3 3v10.5a3 3 0 0 0 3 3h10.5a3 3 0 0 0 3-3V13.5a.75.75 0 0 0-1.5 0v5.25a1.5 1.5 0 0 1-1.5 1.5H5.25a1.5 1.5 0 0 1-1.5-1.5V8.25a1.5 1.5 0 0 1 1.5-1.5h5.25a.75.75 0 0 0 0-1.5H5.25Z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`UI/Controls <WizardHeading {...} /> Should not be able to trigger a click event if it is disabled 1`] = `
<div>
  <div
    class="wizard-heading border-gray-300"
  >
    <div
      class="wizard-heading-circle border-gray-300"
      data-testid="circle-button"
    >
      <svg
        aria-hidden="true"
        class="wizard-heading-icon text-gray-300"
        data-slot="icon"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21.731 2.269a2.625 2.625 0 0 0-3.712 0l-1.157 1.157 3.712 3.712 1.157-1.157a2.625 2.625 0 0 0 0-3.712ZM19.513 8.199l-3.712-3.712-8.4 8.4a5.25 5.25 0 0 0-1.32 2.214l-.8 2.685a.75.75 0 0 0 .933.933l2.685-.8a5.25 5.25 0 0 0 2.214-1.32l8.4-8.4Z"
        />
        <path
          d="M5.25 5.25a3 3 0 0 0-3 3v10.5a3 3 0 0 0 3 3h10.5a3 3 0 0 0 3-3V13.5a.75.75 0 0 0-1.5 0v5.25a1.5 1.5 0 0 1-1.5 1.5H5.25a1.5 1.5 0 0 1-1.5-1.5V8.25a1.5 1.5 0 0 1 1.5-1.5h5.25a.75.75 0 0 0 0-1.5H5.25Z"
        />
      </svg>
    </div>
  </div>
</div>
`;
